File: Dsa-day207.cpp
class Solution {
    private:
    void helper(TreeNode* root,vector<int>&temp){
        if(root == NULL)
        return;

        helper(root->left,temp);
        temp.push_back(root->val);
        helper(root->right,temp);
    }
    TreeNode* build(int s,int e,vector<int>&temp){
        if(s>e)
        return NULL;

        int mid=s+(e-s)/2;

        TreeNode* root=new TreeNode(temp[mid]);
        root->left=build(s,mid-1,temp);
        root->right=build(mid+1,e,temp);

        return root;
    }
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<int>temp;
        helper(root,temp);

        root=build(0,temp.size()-1,temp);
        return root;
    }
};

File: Dsa-day207.py
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        nodes = []
        
        def inorder(node):
            if not node:
                return
            inorder(node.left)
            nodes.append(node)
            inorder(node.right)
        
        def build(left, right):
            if left > right:
                return None
            
            mid = (left + right) // 2
            node = nodes[mid]
            node.left = build(left, mid - 1)
            node.right = build(mid + 1, right)
            return node
        
        inorder(root)
        return build(0, len(nodes) - 1)